From 22753e615fc7a5d136482089c603058cb113f9e6 Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Mon, 22 Feb 2016 15:18:05 +0200
Subject: [PATCH 0082/1345] arm64: exception: add option to ignore
 asynchronous external abort

commit  ad32b22ae2169572ab294a76b67074f40ce5fbcc from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

in certain scenarios failed PCIe configuration accesses during
bootloader phase cause asynchronous external abort which remain
sticky and are triggered only when linux enables external aborts

as a nasty hack, allow linux to ignore such exceptions

Change-Id: I02f1d4ab4bfc3cc2d72afae1dad9b0bb6eaee235
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm64/kernel/entry.S |   10 ++++++++++
 drivers/pci/host/Kconfig  |    9 +++++++++
 2 files changed, 19 insertions(+)

diff --git a/arch/arm64/kernel/entry.S b/arch/arm64/kernel/entry.S
index 93958d1..94ae09f 100644
--- a/arch/arm64/kernel/entry.S
+++ b/arch/arm64/kernel/entry.S
@@ -410,7 +410,11 @@ ENTRY(vectors)
 	kernel_ventry	1, sync				// Synchronous EL1h
 	kernel_ventry	1, irq				// IRQ EL1h
 	kernel_ventry	1, fiq_invalid			// FIQ EL1h
+#ifdef CONFIG_IGNORE_ASYNC_ABORT
+	kernel_ventry	1, ignore
+#else
 	kernel_ventry	1, error_invalid		// Error EL1h
+#endif
 
 	kernel_ventry	0, sync				// Synchronous 64-bit EL0
 	kernel_ventry	0, irq				// IRQ 64-bit EL0
@@ -512,6 +516,12 @@ el1_error_invalid:
 	inv_entry 1, BAD_ERROR
 ENDPROC(el1_error_invalid)
 
+#ifdef CONFIG_IGNORE_ASYNC_ABORT
+el1_ignore:
+	eret
+ENDPROC(el1_ignore)
+#endif
+
 /*
  * EL1 mode handlers.
  */
diff --git a/drivers/pci/host/Kconfig b/drivers/pci/host/Kconfig
index 3e777b8..7e68e38 100644
--- a/drivers/pci/host/Kconfig
+++ b/drivers/pci/host/Kconfig
@@ -7,6 +7,15 @@ config PCI_MVEBU
 	depends on ARM
 	depends on OF
 
+config IGNORE_ASYNC_ABORT
+	bool "Ignore asynchronous abort from configuration access"
+	depends on ARCH_MVEBU
+	help
+	 Ignore asynchronous abort caused during configuration
+	 access to ports with no endpoint connected.
+	 Warning: This will mask all Serror aborts
+	 regardless of their cause
+
 config PCI_AARDVARK
 	bool "Aardvark PCIe controller"
 	depends on ARCH_MVEBU && ARM64
-- 
1.7.9.5

